pandas/core/frame.py

```diff
@@ -637,8 +637,30 @@ class DataFrame(NDFrame, OpsMixin):
     def _constructor(self) -> Callable[..., DataFrame]:
         return DataFrame
 
+    def _constructor_from_mgr(self, mgr, axes):
+        if self._constructor is DataFrame:
+            # we are pandas.DataFrame (or a subclass that doesn't override _constructor)
+            return self._from_mgr(mgr, axes=axes)
+        else:
+            assert axes is mgr.axes
```

Comment:
Is this assert needed here? (it's also not done in `_from_mgr`)

Reply:
Not strictly.  ATM _from_mgr is documented as requiring them to match, so this assertion seemed like an easy way of making it required along this path too.  could remove the assertion and document the requirement in the docstring